File: tools/skill-audit/jassfacts.py
"""war3map.j → 每個 trigger 群組的**機械事實**（⛔ 不做任何技能專屬判斷）。

⭐ 為什麼是一支獨立的模組：三軸驗收裡有兩軸（動畫、傷害）的「原作那一側」
   完全來自這裡，而它的正確性可以**單獨**被驗（不必先跑整個 audit）。

⚠️ 這裡刻意**不吃** `tools/jass-combo/_ground-truth_temp_*.json`：
   那份是 `_temp_` 檔而且 ⛔ **沒有進版控**（`git ls-files` 查過）。
   一份產生的合約文件如果依賴未追蹤的輸入，在別人的機器上就會 `--check` 紅，
   而那正是 2026-08-02「未追蹤來源被烘進產物」那次事故的形狀。
   war3map.j 是**追蹤中的** 2.8MB 原始碼，它才是可以依賴的地面。

⚠️ 這裡也刻意**不猜語意**。它只數得出來的東西：
   `AddSpecialEffect*` 的模型路徑、`UnitDamage*` 的呼叫次數、
   `TriggerSleepAction`/`PolledWait` 的節拍、以及包住它們的 `loop` 有幾圈。
   「這一段是不是一個連段」是 audit.py 的事，不是這裡的事。
"""

from __future__ import annotations

import os
import re
from dataclasses import dataclass, field

# --------------------------------------------------------------------------
# 正則（全部只認**字面**，⛔ 沒有一條認得任何技能的名字）
# --------------------------------------------------------------------------

#: `function Foo takes ... returns ...` —— 函式切割點。
_FUNC = re.compile(r"^function\s+(\w+)\s+takes\b", re.M)
_ENDFUNC = re.compile(r"^endfunction\b", re.M)

#: 四字 rawcode 字面值（`'A07F'`）。⚠️ 單字元的 `'x'` 不是 rawcode，長度必須是 4。
_RAWCODE = re.compile(r"'([A-Za-z0-9_]{4})'")

#: `AddSpecialEffect…( …, "路徑" )` —— 模型路徑一律是呼叫裡的**最後**一個字串。
#: ⚠️ ⛔ 不可以用 `\(([^)]*)\)` 抓參數：`AddSpecialEffectLocBJ( GetUnitLoc(GetTriggerUnit()), "X.mdx" )`
#: 的第一個 `)` 在**巢狀呼叫裡面**，那樣抓會得到空參數 —— 而它的表現是
#: 「這支技能在原作沒有特效」（靜默的假陰性，第二守則失敗形態④）。
#: ⇒ 抓「呼叫起點之後、這一行結束之前」的最後一個字串字面值。BJ 呼叫一律單行。
_SFX = re.compile(r"AddSpecialEffect\w*\s*\(")
_STRLIT = re.compile(r'"((?:[^"\\]|\\.)*)"')

#: 任何 `UnitDamage…` 家族（TargetBJ / PointLoc / …）。
_DAMAGE = re.compile(r"\bUnitDamage\w*\s*\(")

#: 兩種等待。⚠️ `PolledWait` 與 `TriggerSleepAction` 在 WC3 語意不同
#: （前者對多人同步安全），但**對「這支技能有幾拍」這個問題它們是同一件事**。
_WAIT = re.compile(r"\b(?:TriggerSleepAction|PolledWait)\s*\(\s*\(?\s*(-?[\d.]+)")
_WAIT_ANY = re.compile(r"\b(?:TriggerSleepAction|PolledWait)\s*\(")

#: 迴圈邊界：`exitwhen <var> > <上界>`。上界可能是字面值，也可能是一個變數。
_EXITWHEN = re.compile(r"^\s*exitwhen\s+(\S+)\s*>\s*([\w.]+)")
_SETVAR = re.compile(r"^\s*set\s+(\w+)\s*=\s*(-?[\d.]+)\s*$")

#: `InitTrig_<Base>` —— trigger 群組的**註冊表**。群組名一律從這裡來，
#: ⛔ 不是從函式名硬切（`Trig_LinaS_Effect_Actions` 屬於 `LinaS Effect`，不是 `LinaS`）。
_INITTRIG = re.compile(r"^function\s+InitTrig_(\w+)\s+takes\b", re.M)
_INITTRIG_ONE = re.compile(r"^InitTrig_(\w+)$")

#: 週期 timer 驅動的 trigger（世界終結的 12 連擊就住在這種裡面）。
_PERIODIC = re.compile(r"TriggerRegisterTimerEventPeriodic\s*\(\s*\w+\s*,\s*([\d.]+)")

#: 動畫指令。⭐ 這是「動畫正確性」那一軸在原作側的第二個證據來源：
#: 一支技能演出幾個**姿勢**（而不是只有幾拍等待）。
_ANIM = re.compile(r"\bSetUnitAnimation\w*\s*\(")

#: ⭐【移動中的模型特效】原作側的證據（#543 · #551）。
#:
#: 原作把「一具會飛的模型」寫成一隻 locust dummy：`CreateUnit` → 迴圈裡
#: `SetUnitPosition` / `SetUnitX` / `SetUnitY` 一格一格推 → `RemoveUnit`。
#: 所以「這支技能有沒有一具沿路徑移動的模型」在 JASS 裡的痕跡就是**這個計數**
#: （已乘上迴圈圈數）。
#:
#: ⚠️ 它 ⛔ **不等於**「有 dummy」：單獨一次 `SetUnitPosition` 是瞬移／擊退的
#: 落點修正，那是另一件事。判準因此是 `> 1`（見 `audit.py` 的 `move` 軸）——
#: **一段路徑**至少要兩次移動，⛔ 一次不算。
_MOVE = re.compile(r"\b(?:SetUnitPosition|SetUnitX|SetUnitY)\w*\s*\(")

#: ⭐⭐ trigger **鏈**。這是這支工具最重要的一條線。
#:
#: owner 對這張票的原話是「因為有一堆**時間序 JASS + w3x 效果演出**」——
#: 而 WC3 的時間序演出**幾乎不會**寫在同一個 trigger 裡：施法那一個
#: `EnableTrigger` 下一個，下一個再 `EnableTrigger` 下下個（或掛週期 timer）。
#:
#: ⛔ 只用 rawcode 接合會**整段漏掉**：42-04 世界終結的 rawcode `A05D` 只出現在
#: `The_End_ofWorldStart`（1 次傷害），而卡面承諾的「隨機 12 次區域傷害」住在
#: 它 `EnableTrigger` 起來的 `The_End_ofWorldCasting`（0.10 秒週期 timer）裡。
#: 不跟著這條邊走，audit 會對著一支「原作只打 1 下」的技能說 GGD 沒有缺口 ——
#: **一個綠燈，而缺口有 11 下。**
_ENABLE = re.compile(r"\bEnableTrigger\s*\(\s*gg_trg_(\w+)\s*\)")


@dataclass
class JassGroup:
    """一個 trigger 群組（`Trig_<base>_*` 的所有函式）攤平後的可數事實。"""

    base: str
    line: int
    rawcodes: set[str] = field(default_factory=set)
    #: 模型 stem（小寫、去副檔名、去目錄）→ 這個群組呼叫它幾次（已乘上迴圈圈數）
    sfx: dict[str, int] = field(default_factory=dict)
    damage_calls: int = 0
    wait_beats: int = 0
    wait_values: list[float] = field(default_factory=list)
    anim_calls: int = 0
    #: dummy 移動段數（`SetUnitPosition`/`SetUnitX`/`SetUnitY`,已乘上迴圈圈數）
    move_calls: int = 0
    unbounded_loop: bool = False
    #: 這個群組 `EnableTrigger` 起來的下游群組名（見 `_ENABLE` 的註解）
    enables: set[str] = field(default_factory=set)
    #: 它是被週期 timer 驅動的嗎（`TriggerRegisterTimerEventPeriodic`）
    periodic: float | None = None

    def has_facts(self) -> bool:
        """這個群組**演出**了任何東西嗎？

        ⚠️ war3map.j 裡有幾個群組是**表格**而不是演出 —— 例如 `AILearning`
        逐條列出每個英雄的加點順序，於是它一個人就提到了幾百個 rawcode。
        ⛔ 讓它參與 rawcode→技能的接合，會讓**每一支**技能都認領到同一個空群組。
        ⇒ 判準是資料推導的：**一個字都沒演出（沒特效、沒傷害、沒等待、沒動畫、
        沒推動任何一具模型）的群組不算演出**，⛔ 不是「把 AILearning 寫進黑名單」。

        ⚠️ `move_calls` 是 #543 補上的第五項：**把一具模型沿路徑推出去就是演出**，
        而在此之前一個「只推模型、不發特效也不直接扣血」的群組會被整個丟掉 ——
        於是【移動中的模型特效】那一族在原作側**完全隱形**。
        """
        return (
            bool(self.sfx)
            or self.damage_calls > 0
            or self.wait_beats > 0
            or self.anim_calls > 0
            or self.move_calls > 0
        )

    def as_dict(self) -> dict:
        return {
            "base": self.base,
            "line": self.line,
            "rawcodes": sorted(self.rawcodes),
            "sfx": dict(sorted(self.sfx.items())),
            "damageCalls": self.damage_calls,
            "waitBeats": self.wait_beats,
            "waitValues": self.wait_values,
            "animCalls": self.anim_calls,
            "moveCalls": self.move_calls,
            "unboundedLoop": self.unbounded_loop,
            "enables": sorted(self.enables),
            "periodic": self.periodic,
        }


def _stem(path: str) -> str:
    """`"Abilities\\Spells\\Orc\\WarStomp\\WarStompCaster.mdl"` → `warstompcaster`。

    ⭐ 與 `content/assets/vfx/w3x-ability-provenance.json` 的 `stem` 欄位**同一個算法**，
    否則兩邊接不起來（而接不起來會靜默地表現成「這支技能沒有原作美術」）。
    """
    p = path.replace("\\\\", "\\").replace("/", "\\")
    base = p.split("\\")[-1]
    return os.path.splitext(base)[0].strip().lower()


def _loop_multipliers(lines: list[str]) -> tuple[list[int], bool]:
    """替每一行算出「它會被跑幾次」。

    ⚠️ 這一段是動畫軸的**承重**：`Trig_SuperFF7_Actions` 的七連斬長成
    `loop … exitwhen udg_SupI > 7 … endloop`，⛔ 不展開迴圈的話它只有 2 拍，
    展開之後才是 14 拍 —— 而「連斬七次」正是卡面承諾的那個數字。

    上界解析三段式（⛔ 解不出來就誠實承認，不要猜）：
      ① `exitwhen i > 7`      → 字面值 7
      ② `exitwhen i > N` 且前面有 `set N = 12` → 12
      ③ 都不是（來自 `GetUnitAbilityLevel…`）→ 圈數 1，並標記 unbounded
    """
    mult = [1] * len(lines)
    # 先掃出「變數 → 最近一次被指派的字面值」，供 ② 用。
    literal_of: dict[str, float] = {}
    assign_at: list[dict[str, float]] = []
    for ln in lines:
        m = _SETVAR.match(ln)
        if m:
            literal_of[m.group(1)] = float(m.group(2))
        assign_at.append(dict(literal_of))

    stack: list[int] = []          # loop 起始行號
    bounds: dict[int, int] = {}    # loop 起始行號 → 圈數
    unbounded = False
    for i, ln in enumerate(lines):
        s = ln.strip()
        if s == "loop":
            stack.append(i)
            bounds[i] = 1
        elif s == "endloop":
            if stack:
                stack.pop()
        elif stack:
            m = _EXITWHEN.match(ln)
            if m and bounds.get(stack[-1]) == 1:
                raw = m.group(2)
                try:
                    bounds[stack[-1]] = max(1, int(float(raw)))
                except ValueError:
                    val = assign_at[i].get(raw)
                    if val is None:
                        unbounded = True
                    else:
                        bounds[stack[-1]] = max(1, int(val))

    # 第二遍：每一行的乘數 = 包住它的所有 loop 圈數的乘積。
    stack = []
    for i, ln in enumerate(lines):
        s = ln.strip()
        if s == "loop":
            stack.append(i)
            continue
        if s == "endloop":
            if stack:
                stack.pop()
            continue
        m = 1
        for start in stack:
            m *= bounds.get(start, 1)
        mult[i] = m
    return mult, unbounded


def parse_jass(path: str) -> dict[str, JassGroup]:
    """整份 war3map.j → `{群組名: JassGroup}`。"""
    src = open(path, encoding="utf-8", errors="replace").read()
    lines_all = src.split("\n")

    # ① trigger 群組的註冊表。
    bases = sorted({m.group(1) for m in _INITTRIG.finditer(src)}, key=len, reverse=True)

    # ② 切函式。
    funcs: list[tuple[str, int, str]] = []
    starts = [(m.start(), m.group(1)) for m in _FUNC.finditer(src)]
    for idx, (pos, name) in enumerate(starts):
        end = _ENDFUNC.search(src, pos)
        stop = end.end() if end else (starts[idx + 1][0] if idx + 1 < len(starts) else len(src))
        line_no = src.count("\n", 0, pos) + 1
        funcs.append((name, line_no, src[pos:stop]))

    groups: dict[str, JassGroup] = {}

    def group_for(base: str, line_no: int) -> JassGroup:
        g = groups.get(base)
        if g is None:
            g = groups[base] = JassGroup(base=base, line=line_no)
        g.line = min(g.line, line_no)
        return g

    # ① 先掃 `InitTrig_*`：它是「這個 trigger 怎麼被觸發的」的唯一住處。
    #    ⛔ 不可以只掃 `Trig_*_Actions` —— 週期 timer 完全不出現在 Actions 裡。
    for name, line_no, body in funcs:
        m = _INITTRIG_ONE.match(name)
        if not m:
            continue
        g = group_for(m.group(1), line_no)
        p = _PERIODIC.search(body)
        if p:
            g.periodic = float(p.group(1))

    for name, line_no, body in funcs:
        if not name.startswith("Trig_"):
            continue
        # ⭐ **最長**匹配：`Trig_LinaS_Effect_Actions` 要歸 `LinaS_Effect`，⛔ 不是 `LinaS`。
        base = next((b for b in bases if name.startswith("Trig_" + b + "_") or name == "Trig_" + b), None)
        if base is None:
            continue
        g = group_for(base, line_no)

        blines = body.split("\n")
        mult, unbounded = _loop_multipliers(blines)
        g.unbounded_loop = g.unbounded_loop or unbounded

        for i, ln in enumerate(blines):
            k = mult[i]
            for rc in _RAWCODE.findall(ln):
                g.rawcodes.add(rc)
            for m in _SFX.finditer(ln):
                strs = _STRLIT.findall(ln[m.end():])
                if strs:
                    st = _stem(strs[-1])
                    if st:
                        g.sfx[st] = g.sfx.get(st, 0) + k
            g.damage_calls += len(_DAMAGE.findall(ln)) * k
            g.anim_calls += len(_ANIM.findall(ln)) * k
            g.move_calls += len(_MOVE.findall(ln)) * k
            for tgt in _ENABLE.findall(ln):
                if tgt != base:
                    g.enables.add(tgt)
            n_wait = len(_WAIT_ANY.findall(ln))
            if n_wait:
                g.wait_beats += n_wait * k
                for v in _WAIT.findall(ln):
                    g.wait_values.extend([float(v)] * k)
    return groups


def index_by_rawcode(groups: dict[str, JassGroup]) -> dict[str, list[JassGroup]]:
    """rawcode → 提到它的所有群組。

    ⚠️ 一個 rawcode 可能被好幾個群組提到（施法一個、命中效果一個）。
    ⛔ 不要只取第一個 —— 那會讓「命中之後才發生的傷害」整批消失。
    """
    idx: dict[str, list[JassGroup]] = {}
    for g in groups.values():
        if not g.has_facts():
            continue
        for rc in g.rawcodes:
            idx.setdefault(rc, []).append(g)
    for rc in idx:
        idx[rc].sort(key=lambda g: g.line)
    return idx


def closure(groups: dict[str, JassGroup], seeds: list[JassGroup]) -> list[JassGroup]:
    """從施法那一個群組，沿 `EnableTrigger` 邊走到整條演出鏈。

    ⭐ 這是「一支技能在原作到底演了什麼」的正確邊界，⛔ 不是「同名的那一個函式」。
    ⚠️ 去重靠 base 名，而且**排序輸出**（`sim/purity` 那條規矩的精神：
    產生的東西不可以依賴 dict 的走訪順序，否則 `--check` 會隨機紅）。
    """
    seen: dict[str, JassGroup] = {}
    stack = list(seeds)
    while stack:
        g = stack.pop()
        if g.base in seen:
            continue
        seen[g.base] = g
        for nxt in sorted(g.enables):
            child = groups.get(nxt)
            if child is not None and child.base not in seen:
                stack.append(child)
    return [seen[k] for k in sorted(seen)]

File: tools/go/jassfacts/jassfacts.go
// Package jassfacts 是 tools/skill-audit/jassfacts.py 的 Go 逐條移植（票 A3）。
//
// ⭐ 這一支存在的唯一理由是**量一個數字**：owner 2026-08-23 逐字要求
// 「把花最多時間幾支 python 備份後，改用 go 重寫（保留 Python 測試當 reference
// implementation），最後跑相同 test cases 比結果和 benchmark」。
//
// ⚠️⚠️ 讀這支之前請先讀 docs/_reports/A3_temp_20260823b.md 的結論：
// `audit:check` 的 13.98s **不是 Python 慢**，是 jassfacts.py 裡一行
// `src.count("\n", 0, pos)` 的 O(F×N) 二次方掃描（4,719 支函式各自從第 0 位元組
// 重數一次換行 ≈ 6.4 GB）。把它改成增量游標之後 Python 自己就快 11.4×，
// 而**輸出逐位元組相同**。⇒ ⛔ 這支 Go 沒有被接到 package.json 上。
//
// ⛔ 所以這裡的 line_no 用的是**修好之後**的增量算法，⛔ 不是二次方版本 ——
// 拿 Go 去對一個有 bug 的 Python 會把「演算法的勝利」記到「語言」頭上，
// 而那正是這張票要避免的錯誤結論。
package jassfacts

import (
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strconv"
	"strings"
)

// 正則：⛔ 逐條對照 Python 版，一個都不可以「順手改好一點」。
// ⚠️ Go 的 RE2 沒有 backreference，這裡本來就沒用到；但 \b 與 \w 在 Go 是
// **ASCII** 語意，Python 的 str 正則是 **unicode** 語意 —— 兩者只有在
// 「中文字緊貼著 UnitDamage( 這種呼叫」時才會分岔。實測 war3map.j 沒有這種寫法
// （兩邊輸出逐位元組相同），但移植到別份 .j 時這是第一個要重驗的點。
var (
	reFunc       = regexp.MustCompile(`(?m)^function\s+(\w+)\s+takes\b`)
	reEndFunc    = regexp.MustCompile(`(?m)^endfunction\b`)
	reRawcode    = regexp.MustCompile(`'([A-Za-z0-9_]{4})'`)
	reSfx        = regexp.MustCompile(`AddSpecialEffect\w*\s*\(`)
	reStrLit     = regexp.MustCompile(`"((?:[^"\\]|\\.)*)"`)
	reDamage     = regexp.MustCompile(`\bUnitDamage\w*\s*\(`)
	reWait       = regexp.MustCompile(`\b(?:TriggerSleepAction|PolledWait)\s*\(\s*\(?\s*(-?[\d.]+)`)
	reWaitAny    = regexp.MustCompile(`\b(?:TriggerSleepAction|PolledWait)\s*\(`)
	reExitwhen   = regexp.MustCompile(`^\s*exitwhen\s+(\S+)\s*>\s*([\w.]+)`)
	reSetVar     = regexp.MustCompile(`^\s*set\s+(\w+)\s*=\s*(-?[\d.]+)\s*$`)
	reInitTrig   = regexp.MustCompile(`(?m)^function\s+InitTrig_(\w+)\s+takes\b`)
	reInitTrig1  = regexp.MustCompile(`^InitTrig_(\w+)$`)
	rePeriodic   = regexp.MustCompile(`TriggerRegisterTimerEventPeriodic\s*\(\s*\w+\s*,\s*([\d.]+)`)
	reAnim       = regexp.MustCompile(`\bSetUnitAnimation\w*\s*\(`)
	reMove       = regexp.MustCompile(`\b(?:SetUnitPosition|SetUnitX|SetUnitY)\w*\s*\(`)
	reEnable     = regexp.MustCompile(`\bEnableTrigger\s*\(\s*gg_trg_(\w+)\s*\)`)
)

// Group 對應 Python 的 JassGroup。欄位順序＝as_dict() 的鍵順序。
type Group struct {
	Base         string
	Line         int
	Rawcodes     map[string]bool
	Sfx          map[string]int
	DamageCalls  int
	WaitBeats    int
	WaitValues   []float64
	AnimCalls    int
	MoveCalls    int
	UnboundedLoop bool
	Enables      map[string]bool
	Periodic     *float64
}

// HasFacts 對應 Python 的 has_facts()：一個字都沒演出的群組不算演出。
func (g *Group) HasFacts() bool {
	return len(g.Sfx) > 0 || g.DamageCalls > 0 || g.WaitBeats > 0 ||
		g.AnimCalls > 0 || g.MoveCalls > 0
}

// stem 對應 Python 的 _stem()：路徑 → 小寫無副檔名的檔名。
func stem(path string) string {
	p := strings.ReplaceAll(path, `\\`, `\`)
	p = strings.ReplaceAll(p, "/", `\`)
	parts := strings.Split(p, `\`)
	base := parts[len(parts)-1]
	base = strings.TrimSuffix(base, filepath.Ext(base))
	return strings.ToLower(strings.TrimSpace(base))
}

// loopMultipliers 對應 Python 的 _loop_multipliers()。
// ⚠️ 承重點：七連斬寫成 loop … exitwhen i > 7 … endloop，不展開就只有 2 拍。
func loopMultipliers(lines []string) ([]int, bool) {
	mult := make([]int, len(lines))
	for i := range mult {
		mult[i] = 1
	}

	// 「變數 → 最近一次被指派的字面值」逐行快照（供 exitwhen 的第②段用）。
	// ⚠️ Python 每行 dict(literal_of) 複製一份；這裡只在**有變動時**才複製，
	//    語意相同而配置量少一個數量級。
	literalOf := map[string]float64{}
	assignAt := make([]map[string]float64, len(lines))
	for i, ln := range lines {
		if m := reSetVar.FindStringSubmatch(ln); m != nil {
			if v, err := strconv.ParseFloat(m[2], 64); err == nil {
				cp := make(map[string]float64, len(literalOf)+1)
				for k, vv := range literalOf {
					cp[k] = vv
				}
				cp[m[1]] = v
				literalOf = cp
			}
		}
		assignAt[i] = literalOf
	}

	stack := []int{}
	bounds := map[int]int{}
	unbounded := false
	for i, ln := range lines {
		s := strings.TrimSpace(ln)
		switch {
		case s == "loop":
			stack = append(stack, i)
			bounds[i] = 1
		case s == "endloop":
			if len(stack) > 0 {
				stack = stack[:len(stack)-1]
			}
		case len(stack) > 0:
			m := reExitwhen.FindStringSubmatch(ln)
			if m != nil && bounds[stack[len(stack)-1]] == 1 {
				raw := m[2]
				if f, err := strconv.ParseFloat(raw, 64); err == nil {
					bounds[stack[len(stack)-1]] = maxInt(1, int(f))
				} else if val, ok := assignAt[i][raw]; ok {
					bounds[stack[len(stack)-1]] = maxInt(1, int(val))
				} else {
					unbounded = true
				}
			}
		}
	}

	stack = stack[:0]
	for i, ln := range lines {
		s := strings.TrimSpace(ln)
		if s == "loop" {
			stack = append(stack, i)
			continue
		}
		if s == "endloop" {
			if len(stack) > 0 {
				stack = stack[:len(stack)-1]
			}
			continue
		}
		m := 1
		for _, start := range stack {
			if b, ok := bounds[start]; ok {
				m *= b
			}
		}
		mult[i] = m
	}
	return mult, unbounded
}

func maxInt(a, b int) int {
	if a > b {
		return a
	}
	return b
}

type fn struct {
	name string
	line int
	body string
}

// Parse 對應 Python 的 parse_jass()。
func Parse(path string) (map[string]*Group, error) {
	raw, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	src := string(raw)

	// ① trigger 群組註冊表，長度由大到小（⭐ 最長匹配）。
	baseSet := map[string]bool{}
	for _, m := range reInitTrig.FindAllStringSubmatch(src, -1) {
		baseSet[m[1]] = true
	}
	bases := make([]string, 0, len(baseSet))
	for b := range baseSet {
		bases = append(bases, b)
	}
	sort.Slice(bases, func(i, j int) bool {
		if len(bases[i]) != len(bases[j]) {
			return len(bases[i]) > len(bases[j])
		}
		return bases[i] < bases[j] // 同長度不可能同時匹配,排序只為決定論
	})

	// ② 切函式。⭐ line_no 用增量游標（O(N) 全檔一次），⛔ 不是每支從 0 重數。
	locs := reFunc.FindAllStringSubmatchIndex(src, -1)
	funcs := make([]fn, 0, len(locs))
	prevPos, lineNo := 0, 1
	for idx, loc := range locs {
		pos := loc[0]
		name := src[loc[2]:loc[3]]
		stop := len(src)
		if e := reEndFunc.FindStringIndex(src[pos:]); e != nil {
			stop = pos + e[1]
		} else if idx+1 < len(locs) {
			stop = locs[idx+1][0]
		}
		lineNo += strings.Count(src[prevPos:pos], "\n")
		prevPos = pos
		funcs = append(funcs, fn{name: name, line: lineNo, body: src[pos:stop]})
	}

	groups := map[string]*Group{}
	groupFor := func(base string, lineNo int) *Group {
		g, ok := groups[base]
		if !ok {
			g = &Group{
				Base: base, Line: lineNo,
				Rawcodes: map[string]bool{}, Sfx: map[string]int{},
				Enables: map[string]bool{},
			}
			groups[base] = g
		}
		if lineNo < g.Line {
			g.Line = lineNo
		}
		return g
	}

	// ③ 先掃 InitTrig_*（週期 timer 只住在這裡，⛔ 不在 Actions 裡）。
	for _, f := range funcs {
		m := reInitTrig1.FindStringSubmatch(f.name)
		if m == nil {
			continue
		}
		g := groupFor(m[1], f.line)
		if p := rePeriodic.FindStringSubmatch(f.body); p != nil {
			if v, err := strconv.ParseFloat(p[1], 64); err == nil {
				g.Periodic = &v
			}
		}
	}

	// ④ Trig_* 的可數事實。
	for _, f := range funcs {
		if !strings.HasPrefix(f.name, "Trig_") {
			continue
		}
		base := ""
		for _, b := range bases {
			if strings.HasPrefix(f.name, "Trig_"+b+"_") || f.name == "Trig_"+b {
				base = b
				break
			}
		}
		if base == "" {
			continue
		}
		g := groupFor(base, f.line)

		blines := strings.Split(f.body, "\n")
		mult, unbounded := loopMultipliers(blines)
		g.UnboundedLoop = g.UnboundedLoop || unbounded

		for i, ln := range blines {
			k := mult[i]
			for _, m := range reRawcode.FindAllStringSubmatch(ln, -1) {
				g.Rawcodes[m[1]] = true
			}
			for _, m := range reSfx.FindAllStringIndex(ln, -1) {
				strs := reStrLit.FindAllStringSubmatch(ln[m[1]:], -1)
				if len(strs) > 0 {
					if st := stem(strs[len(strs)-1][1]); st != "" {
						g.Sfx[st] += k
					}
				}
			}
			g.DamageCalls += len(reDamage.FindAllStringIndex(ln, -1)) * k
			g.AnimCalls += len(reAnim.FindAllStringIndex(ln, -1)) * k
			g.MoveCalls += len(reMove.FindAllStringIndex(ln, -1)) * k
			for _, m := range reEnable.FindAllStringSubmatch(ln, -1) {
				if m[1] != base {
					g.Enables[m[1]] = true
				}
			}
			nWait := len(reWaitAny.FindAllStringIndex(ln, -1))
			if nWait > 0 {
				g.WaitBeats += nWait * k
				for _, m := range reWait.FindAllStringSubmatch(ln, -1) {
					if v, err := strconv.ParseFloat(m[1], 64); err == nil {
						for j := 0; j < k; j++ {
							g.WaitValues = append(g.WaitValues, v)
						}
					}
				}
			}
		}
	}
	return groups, nil
}
